### PR TITLE
Merging new highlighting stuff

### PR DESCRIPTION
**Pull request by jpolitz**
Making a PR for this so we have a place for discussion.

I fired this up in my local CPO and it's _awesome_.

It looks like it doesn't understand methods in objects – this program doesn't highlight right:

screen shot 2015-11-11 at 2 31 09 pm

Is that an easy fix?

### CODE REVIEW

**Comment by sorawee** [comment]
@jpolitz that's not the right syntax for methods in object, right?

**Comment by blerner** [comment]
I suspect it'll have problems with methods written like that in data definitions too, then...  It's not a trivial fix, I don't think.

**Comment by jpolitz** [comment]
@sorawee I think I didn't typo anything, this runs fine:

```
check:
  { m(self): 5 end }.m() is 5
end
```

**Comment by jpolitz** [comment]
I found this by copy-pasting `anf-loop-compiler.arr` in and seeing what happened.  It did remarkably well for a large chunk of the file.

**Comment by peblair** [comment]
I think that this can be fixed by reversing the function-keyword-to-function-name matching logic. @blerner I assume it would be safe to rely on the Pyret mode to tokenize these things as `function-name` in all cases?
Assuming so, if all goes well, I'll have a fix for this before the weekend. Stay tuned.

EDIT:

Welp, upon further investigation, that won't work as cleanly as I'd hoped. I'll do some more looking into this issue, and, hopefully, I'll still be able to have something by the weekend.

**Comment by jpolitz** [comment]
After the latest update:

screen shot 2015-11-12 at 2 11 23 pm

I think if you can get anf-loop-compiler.arr (https://github.com/brownplt/pyret-lang/blob/master/src/arr/compiler/anf-loop-compiler.arr) to work, you'll have covered all the cases I can think of.

**Comment by blerner** [comment]
Well, _this_ bug at least is very easy to fix -- just treat `for` as an opening keyword (not sure why we forgot to do that).  The method stuff is still hard.

**Comment by jpolitz** [comment]
I think for is an opening keyword.  If I write

```
for 
  5 + 5
end
```

The end snaps to `for`.  I think the issue is that the update makes the `map(...): ... end` pattern be recognized as a method.

**Comment by peblair** [comment]
I have a quick fix\* for that particular issue, but I will need a little bit of time to get some other parts of `anf-loop-compiler.arr` working.

*By "quick fix," I mean "quick fix that doesn't completely fix it in all cases now that I've tested it."

**Comment by jpolitz** [comment]
To be clear – indentation as well as highlighting is incorrect in anf-loop-compiler, I think for the same reason (but maybe not).  Around line 950 things start to get weird (965 with the current version I have):

screen shot 2015-11-15 at 11 41 36 am

File: mode/pyret/pyret.js [hunk 1]
```diff
@@ -370,20 +382,24 @@ CodeMirror.defineMode("pyret", function(config, parserConfig) {
...
+      ls.deferedOpened.fn++;
+      ls.tokens.push("IF");
+      ls.tokens.push("WANTCOLON", "NEEDSOMETHING");
+    } else if (state.lastToken === "else if") {
+      if (hasTop(ls.tokens, "IF")) {
+        if (ls.curOpened.fn > 0) ls.curOpened.fn--;
+        else if (ls.deferedOpened.fn > 0) ls.deferedOpened.fn--;
+        else ls.curClosed.fn++;
         ls.deferedOpened.fn++;
-        ls.tokens.push("IF");
```
Comment (blerner) [inline mode/pyret/pyret.js:377]: I think you still need this line, no?

**Comment by blerner** [comment]
AHHH -- this is because Philip changed how he handles `if`, `else if` and `else` ... and this particular method is named `a-if`, which probably screwed everything up.

**Comment by jpolitz** [comment]
Ahhh interesting.  Keywords as substrings of other names, so tricksy.

**Comment by blerner** [comment]
If you rename that to `a-foobar` in your testing window, does it un-break the indentation?  If so, that's almost definitely the culprit.

**Comment by jpolitz** [comment]
I think it's that there's a field named `block`, which the new system recognizes as a keyword without the `:`.  If I start renaming `block` to `bloc`, things get a little more sane-looking.

**Comment by blerner** [comment]
Ah, yup.  The problem is lines 15--22:

``` js
 const pyret_opening_keywords_colon = ["try", "ask", "ref-graph", "block"];
 const pyret_opening_keywords = pyret_opening_keywords_colon.concat(
                                 ["fun", "when", "for", "if", "let",
                                 "cases", "data", "shared", "check",
                                 "except", "letrec", "lam", "method",
                                 "examples"]);
 const pyret_keywords = 
   wordRegexp(["else if"].concat(pyret_opening_keywords, pyret_closing_keywords,
```

If you trace through this carefully, it means that `block` is a keyword without any colon.  And that does screw up the tokenizer, which then screws everything else up after that.  Philip, `pyret_keywords` and `pyret_keywords_colon` should be disjoint.

**Comment by peblair** [comment]
Well, that was a silly mistake. I'll fix that.

EDIT: Should be fixed now.

**Comment by jpolitz** [comment]
@belph, can you explain the code you added to `codemirror.js` a little bit?  I'm curious if it can be done externally or not.  One thing that would be nice is if we could avoid relying on this fork of the repo at all, and just use CodeMirror proper, along with a few JS files that are hosted within the code.pyret.org repository.  That would avoid having this weird @jpolitz owned repo that things rely on.

**Comment by blerner** [comment]
It looks like it's a monkeypatch to `getTokenAt` that allows returning all the tokens on a line up to a given position, instead of just one the one at the current pos... and it also looks like it's dead code that's never called.

Either way, I _think_ you could define the same thing using `CodeMirror.defineExtension` from an external file.  Probably should, in that case...

**Comment by jpolitz** [comment]
This seems to work well on the compiler file now, except for highlighting.

Here's something that's tempting for now:
1.  Let's use the version without @belph's first "quick fix", which seems to be confused on lots of method-like patterns.
2.  If folks want to use methods, the explicit `method` syntax works just fine; it's only the fancy syntax for methods that has any issue.
3.  In the meantime, we can have a look at that method syntax and see if there's more signposting we want to add in the syntax for it (I can see students being confused by it, for instance), or if we want to write a matching algorithm that correctly tackles it.

**Comment by blerner** [comment]
Joe, can you confirm that in the example you screenshotted above, `e2-visit.block.stmts` **doesn't** highlight `block` as a keyword?  That would be another indication that this is back on track.

**Comment by jpolitz** [comment]
Yeah, it doesn't highlight as a keyword or treat it as something to indent on.

**Comment by jpolitz** [comment]
This does seem to slow down when you get to 500 lines or so.  Editing the end of that compiler file is pretty unusably sluggish.  2048 (about 400 lines), is still OK, though noticeable.

**Comment by blerner** [comment]
Do you notice a difference in performance on opening vs closing keywords?  I'm not sure if the matcher loop is terminating early enough, or if it's unnecessarily scanning the entire file.

Also there's definitely some unnecessary copying going on (e.g. line 549, 550, which copies a copy), which may add some unwanted bloat.

**Comment by jpolitz** [comment]
It was just typing in general; doesn't seem to be tied to the electric keys or anything.  Just writing on a new line seems to be sluggish.

**Comment by blerner** [comment]
@belph I wonder if you could add some performance logging in here, to indicate how many characters `findMatchingToken` needs to process (from which cursor position, to whatever character) each time it gets called.  I'm thinking perhaps it's getting called too often, and especially for non-opening token characters (which are the vast majority of characters) it's scanning way too far...

**Comment by peblair** [comment]
@jpolitz 

> can you explain the code you added to codemirror.js a little bit? 

I can actually roll that file back. `getLineTokens` is a method in new CodeMirror versions which I originally wanted to use, so I added it (it had no other dependencies, from my assessment of the then-current `codemirror.js` file). Since I no longer need it, though, we can likely roll it back.

@blerner 

> I wonder if you could add some performance logging in here

I can (and will) do so, but be aware that a large amount of the sluggishness stems from the matcher's screwing up. For example, I remember seeing a code snippet in `anf-loop-compiler.arr` with something like:

```
for module-name.function-name(stuff):
  <do-things>
```

Highlighting `function-name` erroneously treated the name as a method definition, which read all the way to the end of the file before determining that there were no matching tokens. This is obviously a bug, but it took a solid 5-10 seconds to get that result.

Perhaps @jpolitz's idea of merging in the original code for now would work. If you want to do that, I can go ahead and add some performance metrics to a new branch stemming from that commit. Regardless, in the coming week (or two, I'm fairly swamped this week), I'll hopefully get things at least working, if not performant. I'll also look into seeing if there's some better way to memoize the information for tokens whose open/close regions have not been edited.

**Comment by jpolitz** [comment]
Thanks for the update.

I think that it's worth considering skipping syntax that makes it this hard to write a matcher.  In the future folks will want to contribute Sublime Text and Vim and Atom and GEdit and who-knows-what-editor modes, and this will be a sticking point for them, too.

**Comment by blerner** [comment]
@belph Ah, good, that's intuitively reasonable.  I was afraid there was an inadvertent O(n^2) algorithm involved, but I'm glad it's just overly-permissive matcher guesses.  There's two things we can do here:
- Instead of immediately calling `findMatchingToken` every single time "cursorActivity" fires, set a timeout for a second from now, and do it then.  If the typist is paused for a second, then pausing another second to wait for matching shouldn't be too onerous.
- Hard-code reasonable limits beyond which it's pointless to search every single time.  (I mean, it's not terribly useful if the match is wildly off-screen...)  Though we may want to crib emacs/DrRacket's ability to go-to-matching-delimiter, and for that I'd be willing to wait for an accurate check.

Later, I think the right thing to do is to have the actual indenter apply marks to text to notate which names are really method definitions, for use by the highlighter.  

@jpolitz Yup, possibly.  Our shorthand method syntax (both in object literals and in data definitions) lacks a clear opening token to match its end token, so that's a drag, even if it's beautifully concise.  Also unfortunately, if we're going to talk about editor support, triple-quoted strings are very painful to support well, because most editors I know of only support two-character delimiters (`/* thanks C! */`).  We may need to revisit that.  And finally, we might also want to have the highlighter handle `#| block comment delimiters |#`, or at least that go-to-matching-delimiter...

**Comment by jpolitz** [comment]
I'm a little less worried about triple-backtick, because most editors will support Python well, and we can steal whatever those modes use for `"""` in Python.  But yeah, it is a little odd.  Maybe double-backtick?

Snapping block comments is a good idea.

**Comment by blerner** [comment]
From experience looking at the python mode for emacs, no, triple-quote isn't supported well; it's a seriously ugly hack ;-)  I tried stealing that for Pyret's mode, and it doesn't work reliably at all.  

On second thought, I don't think we want block comments to have to be well-nested with regard to other delimiters (commenting out unbalanced parens or function headers doesn't seem too outlandish to me), so I don't want to add support for that into the highlighter.  Just jumping from opener token to closer token, once we add that...

**Comment by jpolitz** [comment]
@belph, is 6b20da1ba51ec822d7987ab848de5bedeb2e88a5 the main commit to revert in order to back out the method-matching behavior?  I reverted that, and it doesn't seem to break indentation at all, just doesn't match well.  I think performance is the only thing I'd worry about for putting it out there at this point.

**Comment by peblair** [comment]
@jpolitz It should be; however, is it possible to still keep the latest
commit even if you don't see any weird indentation? The bug that f21b4f2
fixes _should_ still be present in that version as well, I'm guessing (not
at my machine right now, so I can't check). As such, it is probably a good
idea to keep it around, so long as it doesn't break anything (which it
really shouldn't). My git-fu isn't good enough to know for sure, but I
think what I'm asking is done with `git cherry-pick` (although I'm sure you
two know how to do it :) ).

**Comment by jpolitz** [comment]
Yeah I just did `git revert 6bc71fc8cd71af6514657b964a95026addfde256` which reverses only the changes in that commit.

**Comment by jpolitz** [comment]
Actually, @belph, can you run that revert command (and push)?  Then we can merge and close this PR, I think.

**Comment by peblair** [comment]
Alright, @jpolitz, it's done. FYI I botched the reversion initially, so on the off-chance you pulled within the last 10min, you should reset to the repository's state (it was enough of a screw-up that I was did the evil correction tactic of `git push --force`)

**Comment by jpolitz** [comment]
Git is so friendly.

**Comment by jpolitz** [comment]
Should this make code folding work?  Like if I turn on foldGutters in CodeMirror, should this do anything?

**Comment by peblair** [comment]
Yes, actually. That's one of the reasons I went with this approach :). I have that enabled on my local copy with the following in `cmOptions` inside of `/src/web/js/repl-ui.js` (Line 38):

```
"Ctrl-Q": function(cm) { cm.foldCode(cm.getCursor()); }
```

Any region which matches in the highlighter (parens, keywords, etc.) can be folded with this.

**Pull request merged**